packages/next/src/lib/metadata/resolvers/resolve-url.test.ts

```diff
@@ -106,3 +110,71 @@ describe('resolveAbsoluteUrlWithPathname', () => {
     })
   })
 })
+
+describe('getSocialImageFallbackMetadataBase', () => {
+  describe('fallbackMetadataBase', () => {
+    let originalEnv: NodeJS.ProcessEnv
+    function getSocialImageFallbackMetadataBaseHelper(): string {
+      return getSocialImageFallbackMetadataBase(null).fallbackMetadataBase.href
+    }
+
+    beforeEach(() => {
+      originalEnv = process.env
+    })
+
+    afterEach(() => {
+      delete process.env.VERCEL_URL
+      delete process.env.VERCEL_ENV
+      delete process.env.VERCEL_BRANCH_URL
+      delete process.env.VERCEL_PROJECT_PRODUCTION_URL
+
+      process.env = originalEnv
+    })
+
+    it('should return localhost url in local dev mode', () => {
+      // @ts-expect-error override process env
+      process.env.NODE_ENV = 'development'
+      expect(getSocialImageFallbackMetadataBaseHelper()).toBe(
+        'http://localhost:3000/'
+      )
+    })
+
+    it('should return metadataBase ', () => {
+      // @ts-expect-error override process env
+      process.env.NODE_ENV = 'production'
+      expect(getSocialImageFallbackMetadataBaseHelper()).toBe(
+        'http://localhost:3000/'
+      )
+    })
+
+    it('should prefer branch url in preview deployment if presents', () => {
+      // @ts-expect-error override process env
+      process.env.NODE_ENV = 'production'
+      process.env.VERCEL_ENV = 'preview'
+      process.env.VERCEL_BRANCH_URL = 'branch-url'
+      process.env.VERCEL_URL = 'vercel-url'
+      expect(getSocialImageFallbackMetadataBaseHelper()).toBe(
+        'https://branch-url/'
+      )
+    })
+
+    it('should return vercel url in preview deployment if only it presents', () => {
+      // @ts-expect-error override process env
+      process.env.NODE_ENV = 'production'
+      process.env.VERCEL_ENV = 'preview'
+      process.env.VERCEL_URL = 'vercel-url'
+      expect(getSocialImageFallbackMetadataBaseHelper()).toBe(
+        'https://vercel-url/'
+      )
+    })
+
+    it('should return project production url in production deployment', () => {
+      // @ts-expect-error override process env
+      process.env.NODE_ENV = 'production'
+      process.env.VERCEL_PROJECT_PRODUCTION_URL = 'production-url'
+      expect(getSocialImageFallbackMetadataBaseHelper()).toBe(
+        'https://production-url/'
+      )
+    })
+  })
```

Comment:
Probably add one more test that has both `VERCEL_URL` and `VERCEL_PROJECT_PRODUCTION_URL` to make sure the correct one is selected